alpakka/src/main/scala/org/scanamo/RetryUtility.scala

```diff
@@ -0,0 +1,43 @@
+package org.scanamo
+
+import java.util.concurrent._
+
+import com.amazonaws.services.dynamodbv2.model._
+
+import scala.concurrent.duration._
+import scala.concurrent.{Future, Promise, _}
+
+object RetryUtility {
+
+  private final val scheduler: ScheduledExecutorService = Executors.newScheduledThreadPool(1)
+
+  def retryWithBackOff[T](op: => Future[T],
+                          retrySettings: RetrySettings)(implicit executionContext: ExecutionContext): Future[T] =
+    op.recoverWith {
+      case exception @ (_: InternalServerErrorException | _: ItemCollectionSizeLimitExceededException |
+          _: LimitExceededException | _: ProvisionedThroughputExceededException | _: RequestLimitExceededException) =>
+        val retries = retrySettings.retries
+        val initialDelay = retrySettings.initialDelay.toMillis
+        val factor = retrySettings.factor
+        if (retries > 0) {
+          for {
+            _ <- waitForMillis(initialDelay)
```

Comment:
Same here, let's move that to `RetrySettings`:
```scala
trait RetrySettings {
  def delay: Long
}
```

Now i think `RetrySettings` should be named `RetryPolicy`.

Reply:
@regiskuckaertz initialDelay is already coming from RetrySettings.
Can you please elaborate your suggestion?

Reply:
The waiting time (`initialDelay` or what I call `delay` above) is a function of the particular policy:

- a constant policy will always yield the same amount
- a linear one will yield `n * d`
- an exponential one will yield `n ^ f * d`

Right now you have implemented the second, but we'll add more and I think it's best if the loop that you wrote doesn't concern itself with the details of the calculation. How about this:

```scala
sealed trait RetryPolicy { self =>
  final val done: Boolean = self match {
    case Max(0)    => true
    case And(x, y) => x.done && y.done
    case Or(x, y)  => x.done || y.done
    case _         => false
  }

  final val delay: Long = self match {
    case Constant(d)          => d.toMillis
    case Linear(d, n)         => d.toMillis * n
    case Exponential(d, n, f) => d.toMillis * Math.pow(n, f)
    case And(x, y)            => Math.max(x.delay, y.delay)
    case Or(x, y)             => Math.min(x.delay, y.delay)
    case _                    => 0
  }

  final def update: RetryPolicy = self match {
    case x: Constant    => x
    case x: Linear      => x.copy(n = n + 1)
    case x: Exponential => x.copy(n = n + 1)
    case Max(m)         => Max(m - 1)
    case And(x, y)      => And(x.update, y.update)
    case Or(x, y)       => Or(x.update, y.update)
  }

  final def &&(that: RetryPolicy): RetryPolicy = And(self, that)
  final def ||(that: RetryPolicy): RetryPolicy = Or(self, that)
}

object RetryPolicy {
  private[RetryPolicy] final case class Constant(d: FiniteDuration) extends RetryPolicy
  private[RetryPolicy] final case class Linear(d: FiniteDuration, n: Int) extends RetryPolicy
  private[RetryPolicy] final case class Exponential(d: FiniteDuration, n: Int, f: Double) extends RetryPolicy
  private[RetryPolicy] final case class Max(m: Int) extends RetryPolicy
  private[RetryPolicy] final case class And(x: RetryPolicy, y: RetryPolicy) extends RetryPolicy
  private[RetryPolicy] final case class Or(x: RetryPolicy, y: RetryPolicy) extends RetryPolicy
}
```

With that we can express many flavours of policies, with the addition perhaps of two constructors: `Always` and `Never`.

Reply:
I think we also want a `Capped(m: Long, p: RetryPolicy)` that will use `p.delay` but cap at `m`.

Reply:
@regiskuckaertz If I understand you correctly, As I added the number of retried then there is no need to add always & never cap too.

Moreover, I added the number of policies that you suggest to add in it.
Please take a look at it